### PR TITLE
Gregsdennis/format registry

### PR DESCRIPTION
**Pull request by gregsdennis**
### What kind of change does this PR introduce?

Feature

### Issue & Discussion References

Related to #1552

### Summary

- Adds a `format` registry file and includes OpenAPI 3.1.1 formats.  In the issue, I mocked up the registry with an array at the root, but realized an object would be good since we really shouldn't have multiple definitions for the same format.  (I have other thoughts on this that I'll post in the issue and update here if necessary.)
- Adds text defining the registry and requirements for implementations to support them.  Having a link to a later section feels weird to me.  Maybe it's not?  Happy to reorganize if others think it's off.

### Does this PR introduce a breaking change?

No.  There is a new "SHOULD" requirement to support the registry formats.  I'd like this to be _slightly_ less required than the ones defined by JSON Schema, but "MAY" seems not required enough.  Open to entertaining ideas.

### CODE REVIEW

**Comment by ralfhandl** [comment]
@gregsdennis What about the formats registered in https://spec.openapis.org/registry/format/index.html?
Will they be eventually registered here, too?

And what will be the process/hurdles for registering formats?

Goal would be to retire the OAS format registry and instead redirect here.

**Comment by gregsdennis** [comment]
I didn't know that page existed.  I was just looking at your spec. 

Yeah, I expect those can be pulled in as well. The plan _was_ to transfer your registry to us.

Given the sheer volume of formats, I'm more inclined to use "MAY" for the registry formats, but also reserve those names so that they can only mean what's registered.  That way _if_ those names are supported, they MUST be these formats.

**Comment by karenetheridge** [comment]
We'll want to support an array for hyperlinks that reference the specification, as I think we'll want to specify multiple versions of reference documents. For example, it's useful to note the earliest version of the OAS spec that added mention of a format, but there are also cases where the reference specification for a format has changed over time (e.g. for json-schema's hostname format, draft7 uses RFC1034, draft2019-09+ uses RFC1123 and that information shouldn't be hidden by simply linking to draft7 and draft2019-09. Perhaps we could say something like "RFC1034 via draft7, RFC1123 via draft2019-09 and draft2020-12", but I'm not sure how that could be visibly represented in a compact yet readable way.

The formats listed here so far are present in OpenAPI v3.1.1, but we should also start out by listing all the formats available in various drafts of JSON Schema itself.

Also note that the current OAS format registry is at https://spec.openapis.org/registry/format/index.html and we should include the information contained there if we agree to take on hosting of that registry, although I've spotted a few errors in that data so those will need to be fixed.

**Comment by karenetheridge** [comment]
> I didn't know that page existed. I was just looking at your spec.

I'm confused, we talked about proposing a transfer of the registry several months ago here?  https://github.com/json-schema-org/json-schema-spec/issues/1552#issuecomment-2449071062

**Comment by gregsdennis** [comment]
I've imported all of the formats.  It _is_ a direct import processed by GH Copilot.  Checking on the errors...

**Comment by gregsdennis** [comment]
> I'm confused... @karenetheridge

The answer is in your remark: "several months ago" 😆

**Review by ralfhandl — APPROVED** [review]

**Comment by gregsdennis** [comment]
@json-schema-org/tsc I think thisi is one that we can pull in and even publish before we release the spec.

There's still a fair amount of planning, design, and work to do to get it on the website, but this is a critical step for that to take place.

File: specs/registries/format.json [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+        "examples": ["token"],
+        "deprecated": false,
+        "supportedBy": []
+    },
+    "time": {
+        "description": "Time as defined by full-time - RFC3339",
+        "definingBody": "JSON Schema",
+        "definition": "https://json-schema.org/draft/2020-12/json-schema-validation.html#name-dates-times-and-duration",
+        "types": ["string"],
+        "examples": ["13:45:30"],
```
Comment (GREsau) [inline specs/registries/format.json:326]: https://www.rfc-editor.org/rfc/rfc3339.html#section-5.6 defines `full-time` as:
```
time-numoffset  = ("+" / "-") time-hour ":" time-minute
time-offset     = "Z" / time-numoffset

full-time       = partial-time time-offset
```

`time-offset` is non-optional and non-empty, so I don't think `"13:45:30"` is valid in this scenario - it would need to be something like `"13:45:30Z"` or `"13:45:30+01:00"`

File: specs/jsonschema-validation.md [hunk 2]
```diff
@@ -328,10 +328,10 @@ an annotation.
 
 Implementations:
 
-- SHOULD provide validation for each format attribute defined in this
-  document;
-- MAY support format values not defined in this document, but such support MUST
-  be configurable and disabled by default;
+- SHOULD provide validation for each format attribute defined in this document
+  and listed in the {{format-registry}};
```
Comment (jdesrosiers) [inline specs/jsonschema-validation.md:332]: I can see the registry growing to have hundreds of entries. I don't think suggesting that implementations should implement all of them is realistic.

File: specs/jsonschema-validation.md [hunk 3]
```diff
@@ -328,10 +328,10 @@ an annotation.
...
+- MAY support format values not defined in this document or listed in the
+  registry, but such support MUST be configurable and disabled by default;
 - SHOULD use a common parsing library or a well-known regular expression for
   each format;
 - SHOULD clearly document how and to what degree each format attribute is
```
Comment (jdesrosiers) [inline specs/jsonschema-validation.md:337]: I don't think this bullet point makes sense. We don't allow partial validation anymore, so full validation is the only valid way to validate a format.

File: specs/jsonschema-validation.md (hunk 2 above)
Comment (gregsdennis) [reply specs/jsonschema-validation.md:332]: We need a middle ground between SHOULD and MAY for the registry.  Kind of a "hey it's a good idea if you want to, but you don't, no big deal."

File: specs/jsonschema-validation.md (hunk 3 above)
Comment (gregsdennis) [reply specs/jsonschema-validation.md:337]: I agree, but realistically implementations are going to support formats only as far as they care to, even if we include a bunch of edge-case tests.  I think having a note that requires implementations to declare their support (or maybe their limitations) covers that base.

File: specs/jsonschema-validation.md (hunk 2 above)
Comment (GREsau) [reply specs/jsonschema-validation.md:334]: "Implementations are encouraged to X" is a pattern I've noticed in a few RFCs, is that the type of thing you're looking for?

e.g.
```
- SHOULD provide validation for each format attribute defined in this document;
- are encouraged to provide validation for format attributes listed in the {{format-registry}};
```

Comment (jdesrosiers) [reply specs/jsonschema-validation.md:334]: I like "encouraged". How about something like, "implementations are encouraged implement formats from the registry and SHOULD (MUST?) provide an extension mechanism for formats they don't support".

The idea is that you either need to provide support for every format in the registry or provide a mechanism for the user to plug in support for anything that's missing.

File: specs/jsonschema-validation.md (hunk 3 above)
Comment (jdesrosiers) [reply specs/jsonschema-validation.md:339]: I see what you're saying, but I feel like it gives permission for incomplete validation, which we don't want.

> (or maybe their limitations) 

Yeah, that. I think that's better. You shouldn't have to say you support what you're required to support. You should only have declare when your support is lacking in some way.

File: specs/jsonschema-validation.md (hunk 2 above)
Comment (gregsdennis) [reply specs/jsonschema-validation.md]: Updated

File: specs/jsonschema-validation.md (hunk 3 above)
Comment (gregsdennis) [reply specs/jsonschema-validation.md]: Updated

**Review by jdesrosiers — APPROVED** [review]
